### ISSUE TITLE
[BUG] inout parameterized statement generates wrong Cpp1

### ISSUE THREAD

**Issue by ntrel** [hsutter/cppfront#393]
```cpp
    (inout i := 0) while i < 4 next i++ {}
```
cppfront compiles to Cpp1 fine, but Cpp1 declares the `i` variable wrongly:
```
$ g++ -Wall -o "test" "test.cpp" -I/home/nick/git/cppfront/include --std=c++2b
test.cpp: In function ‘int main()’:
test.cpp:52:11: error: cannot bind non-const lvalue reference of type ‘int&’ to an rvalue of type ‘int’
   52 | auto& i = 0;
      |           ^
```

**Comment by JohelEGP** [comment]
It's not wrong. I think you're expected to give it a valid initializer just as you would an equally declared parameter.

**Comment by JohelEGP** [comment]
For this case, you might want to "pass (the initializer) by" `copy`: <https://godbolt.org/z/dPMvP8q1j>.
```Cpp2
main: () = {
    (copy i := 0) while i < 4 next i++ {}
}
```
Though that hits:
```output
main.cpp2(2,19): error: (temporary alpha limitation) parameters scoped to a block/statement must be 'in' (the default) or 'inout'
```
`copy` isn't one of the *parameter-direction*s argued against at <https://github.com/hsutter/cppfront/issues/386#issuecomment-1519109176>.

**Comment by JohelEGP** [comment]
Another option is
making `copy` independent of `in`
by removing its `const`-by-default.
So that
`x` in `:(copy x: int) = { }` has type `int` and
`y` in `:(copy x: const int) = { }` has type `const int`.

That seems like moving away from `const`-by-default.
But it also increases the expressiveness of *parameter-direction*s.
It'd become possible to write coroutines and functions to `std::thread`
with pass-by-value parameters that can be modified.

The default _parameter-direction_ for a parameterized statement could be `copy`.

**Comment by msadeqhe** [comment]
I think this is the grammar of parameterized statement blocks as described in [this comment](https://github.com/hsutter/cppfront/issues/391#issuecomment-1519374079):

> It's obvious enough that if `(...)` is for initialization or for argument passing in parameterized statement blocks, the syntax will be generally like this:
> * In the following list: `construct` may be `if`, `while`, `do`, `for`, `next` or anything else in the future. `(inits)` is for initialization and `(params)` is for argument passing in parameterized statement blocks.
> * `(inits) { ... }`
> * `(inits) construct { ... }`
> * `(inits) construct (params) { ... }`
> * `construct (params) { ... }`
> * In a nutshell, `(params)` are after `construct`, and `(inits)` are at the start of control structures. It is a simple rule.

`(inits)` part is a syntax suger for declaring local variables in nested scope, they may be `inout` by default just like how local variables are not `const` by default. On the other hand, they may be `in` by default just like how function parameters are `in` by default.

But `(params)` part is a little different, arguments are passed indirectly, therefore they may be `in` by default as they are now.

---

> Reading [#386 (comment)](https://github.com/hsutter/cppfront/issues/386#issuecomment-1520146455), the issue seems to be that [#386 (comment)](https://github.com/hsutter/cppfront/issues/386#issuecomment-1518194517)'s examples don't really work:
> 
> > _Note Because this works naturally for flow control statements, those no longer need a special syntax for `if`/`switch`/etc. scoped variables. For example:_
> > ```
> > (x:int = f()) if x>1		// same as C++17: if (int x = f(); x>1)
> > (x:int = f()) switch x 		// same as C++17: switch (int x = f(); x)
> > (x:int = f()) for x>1 next --x	// same as K&R C: for (int x = f(); x>1; --x)
> > (x:int = f()) for range do	// same as C++20: for (int x = f(); _ : range)
> > (x:int = f(), i:=0) for range do next ++i
> >                                 // same as C++20: { int i=0; for (int x = f(); _ : range) { ++i;
> > (x:int = f()) while x>1		// not yet allowed for ‘while’ in C++20
> > (x:int = f()) do { } while x>1	// not yet allowed for ‘do’ in C++20
> > (x:int = f()) try			// not yet allowed for ‘try’ in C++20
> > 				// note x is available inside the catch too
> > ```
> 
> I, too, expected a parametrized statement of a single parameter to subsume all of Cpp1's `if` and loops variable declaration syntax.

Yes, good point. It seems the design is changed a little bit. Local variables cannot be references (e.g. `&` or `&&` in Cpp1). But function parameters can be either `in`, `inout`, `out`, `copy`, `move` or `forward`. Both of them can hold pointer types. Therefore `(inits)` part of parameterized statement blocks are different from initialization part of Cpp1 control structures.

**Comment by JohelEGP** [comment]
> As an optimization of `in`, I think `copy` would still declare the variable as `const`. I don't think there's a _parameter-direction_ to simulate the variable declaration `i := 0;`.

Reading https://github.com/hsutter/cppfront/commit/8d2f274050b7984b55a4cc46d3104892c9048755#diff-43ba71198da8225f205041c699dba495883d7d416c0b58916bf226aae97bafaeR3196-R3198 made me realize that this actually works: <https://godbolt.org/z/z7W1q5ohW>.
```Cpp2
next: (copy x: int) -> int = x++;
main: () -> int = next(0);
```
So I think the answer is "use the `copy` _parameter-direction_".

**Comment by JohelEGP** [comment]
as suggested above at https://github.com/hsutter/cppfront/issues/393#issuecomment-1520170739.
It should be as easy as adding
`&& param->direction() != passing_style::copy`
and updating the error message at https://github.com/hsutter/cppfront/commit/928186e4db6f7c3fba7b8df8b4ff9648e1c938e9#diff-43ba71198da8225f205041c699dba495883d7d416c0b58916bf226aae97bafaeR5124-R5127.

**Comment by msadeqhe** [comment]
If I understand your suggestion correctly, in this case `copy` has overhead for most user-defined types.

**Comment by msadeqhe** [comment]
> Reading [8d2f274#diff-43ba71198da8225f205041c699dba495883d7d416c0b58916bf226aae97bafaeR3196-R3198](https://github.com/hsutter/cppfront/commit/8d2f274050b7984b55a4cc46d3104892c9048755#diff-43ba71198da8225f205041c699dba495883d7d416c0b58916bf226aae97bafaeR3196-R3198) made me realize that this actually works: https://godbolt.org/z/z7W1q5ohW.

You're right. I hadn't read that comment from that commit. So  I get your point now. You mean the behaviour of `copy` parameter type is similar to local variable declaration. That's smart to be the default parameter type.

**Comment by AbhinavK00** [comment]
Question: [8d2f274](https://github.com/hsutter/cppfront/commit/8d2f274050b7984b55a4cc46d3104892c9048755#diff-43ba71198da8225f205041c699dba495883d7d416c0b58916bf226aae97bafaeR3196-R3198) mentions that `copy` parameters can be const-qualified. May I ask what is the use of `const copy` parameters as one should use `in` parameter in that case. Are there any use cases where `in` would be insufficient?

Another Question: This one is probably for Herb, how does he plan to handle `volatile` in cpp2, the same way as cpp or some other new way. I think there was a paper by JF Bastein about deprecating `volatile`, would Herb like to go that direction and implement `volatile_load` and `volatiles_store` instead?

**Comment by JohelEGP** [comment]
It's likely an error to give a coroutine or `std::thread`'s function a reference argument. See https://github.com/hsutter/cppfront/issues/393#issuecomment-1520593893, https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#SScp-coro, https://github.com/hsutter/cppfront/issues/76#issuecomment-1279502156.

**Comment by hsutter** [comment]
Thanks! Added `copy` local parameters.

**Comment by hsutter** [comment]
> This one is probably for Herb, how does he plan to handle `volatile` in cpp2

Actually primarily by a simple rename... as a `cpp2::alien_memory<T>` type (or pseudotype), but no cv-qualifiers or similar support in the language.

**Comment by AbhinavK00** [comment]
That's great, I was hoping for something along those lines and your simple idea is even better than something I had in mind

**Comment by hsutter** [comment]
@AbhinavK00 Sure: When you use `copy` you are asking for a new variable. That variable may be `const` or not. For a function parameter, that matters in the function body (not to the caller). For a statement/block parameter, the same is true; non-`const` is a scratch variable that is local to the statement/block that the body can modify, `const` is one that it can't modify.

**Comment by AbhinavK00** [comment]
Ah ok, I was thinking only for functions and there it doesn't matter to caller so I couldn't see any use but it may make sense for block parameters which I currently have to wrap my head around.
